File: python/techai_webutils/src/techai_webutils/core/interfaces/event_publisher.py
"""Event publisher interface for domain events.

Mirrors Go's ``interfaces.EventPublisher``.
"""

from __future__ import annotations

from abc import ABC, abstractmethod

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from techai_webutils.core.events.events import DomainEvent


class EventPublisher(ABC):
    """Publishes domain events, typically via a transactional outbox."""

    @abstractmethod
    async def publish(self, event: DomainEvent) -> None:
        """Publish a single domain event."""
        ...

    @abstractmethod
    async def publish_batch(self, events: list[DomainEvent]) -> None:
        """Publish multiple domain events."""
        ...

File: go/clients/jobs/river/event_publisher.go
package river

import (
	"context"

	"github.com/gt-tech-ai/knowledge-engine/go/core/events"
	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
)

// Compile-time interface assertion.
var _ interfaces.EventPublisher = (*EventPublisher)(nil)

// EventPublisher is a no-op EventPublisher: it accepts events without publishing
// them (a transactional outbox through River is not implemented).
type EventPublisher struct {
	// enqueuer is the job enqueuer used for transactional event publishing.
	enqueuer interfaces.JobEnqueuer
}

// NewEventPublisher creates a new event publisher.
func NewEventPublisher(enqueuer interfaces.JobEnqueuer) *EventPublisher {
	return &EventPublisher{
		enqueuer: enqueuer,
	}
}

// Publish accepts event without publishing it and returns nil.
func (p *EventPublisher) Publish(ctx context.Context, event events.Event) error {
	return nil
}

// PublishBatch calls Publish for each event, stopping at the first error.
func (p *EventPublisher) PublishBatch(ctx context.Context, evts []events.Event) error {
	for _, event := range evts {
		if err := p.Publish(ctx, event); err != nil {
			return err
		}
	}
	return nil
}
